File: Data Structures & Algorithms/permutations/submission-2.cpp
class Solution {
public:
    vector <vector<int>> res;
    vector <int> arr;
    void backtrack(vector<int>& nums, vector<bool> used){
        if( arr.size() == nums.size()) {
            res.push_back(arr);
            return;
        }
        for (int i = 0; i < nums.size();i++){
            if(!used[i]){
                arr.push_back(nums[i]);
                used[i] = true;
            
            backtrack(nums, used);
            used[i] = false;
            arr.pop_back();}

        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<bool> used(nums.size(), false);
        backtrack(nums, used);
        return res;
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.java
class Solution {
    int currSum = 0;
    List<Integer> curr = new ArrayList<>();
    List<List<Integer>> res = new ArrayList<>();

    public void backtrack(int[]nums, int start, int target){
       if(currSum == target){
        res.add(new ArrayList<>(curr));
        return;
       }
       for (int i = start; i < nums.length; i ++){
        if (currSum+ nums[i] > target) break;
        if(i >start && nums[i] == nums[i-1]) continue;
        currSum+= nums[i];
        curr.add(nums[i]);
        backtrack(nums, i+1, target);
        curr.remove(curr.size()-1);
        currSum -= nums[i];
       }
    }
          public List<List<Integer>> combinationSum2(int[] candidates, int target) {
            Arrays.sort(candidates);
            backtrack(candidates, 0, target);
            return res;
        
    }
}
